llvm/lib/AsmParser/LLParser.cpp

```diff
@@ -3981,6 +3981,33 @@ bool LLParser::parseValID(ValID &ID, PerFunctionState *PFS, Type *ExpectedTy) {
     return false;
   }
 
+  case lltok::kw_splat: {
+    Lex.Lex();
+    if (parseToken(lltok::lparen, "expected '(' after vector splat"))
+      return true;
+    Constant *C;
+    if (parseGlobalTypeAndValue(C))
+      return true;
+    if (parseToken(lltok::rparen, "expected ')' at end of vector splat"))
+      return true;
+
+    if (auto *CI = dyn_cast<ConstantInt>(C)) {
```

Comment:
Why is the special ConstantInt/ConstantFP handling here needed?

Reply:
Constant values are parsed separately from the type of the resulting expression, with the `ValID` union used as a proxy to hold the parsed value.

Reply:
It turns out when extending the work to cover non-immediate values I inadvertently introduced a parsing bug so I don't think I can safely use `t_Constant`.  That said, this doesn't effect the core of your suggestion so I'll funnel all the splat cases down the same path, I'll just need to give it a new name.